File: Qb.Js/src/Logic.ts
import { WhereClause } from "./WhereClause";
import { Cond } from "./Cond";
import { WhereRel } from "./Enums";
import { Utils } from "./Utils";



export class Logic {
  Clause: WhereClause;

  static And(conds: Cond[]): Logic;
  static And(logics: Logic[]): Logic;
  static And(...logics: Logic[]): Logic;
  static And(...opers: Cond[]): Logic;
  static And(): Logic {
    if (arguments.length == 0)
      return Logic.AndOrLogic(WhereRel.And, new Array<Logic>());

    let arr;
    if (arguments[0] instanceof Array)
      arr = arguments[0];
    else
      arr = arguments;

      if (arr.length > 0) {
        let isCond = arr[0] instanceof Cond;
        if (isCond) {
          let c = new Array<Cond>();
          for (let i = 0; i < arr.length; i++) {
            c.push(arr[i]);
          }
          return Logic.AndOrOper(WhereRel.And, c);
        } else {
          let l = new Array<Logic>();
          for (let i = 0; i < arr.length; i++) {
            l.push(arr[i]);
          }
          return Logic.AndOrLogic(WhereRel.And, l);
        }
      }
  }

  static Or(...logics: Logic[]): Logic;
  static Or(...opers: Cond[]): Logic;
  static Or(): Logic {
    if (arguments.length >= 1) {
      if (arguments[0] instanceof Logic) {
        var logics = new Array<Logic>();
        for (var i = 0; i < arguments.length; i++) {
          logics.push(arguments[i]);
        }
        return Logic.AndOrLogic(WhereRel.Or, logics);
      }
    }
    var opers = new Array<Cond>();
    for (var i = 0; i < arguments.length; i++) {
      opers.push(arguments[i]);
    }
    return Logic.AndOrOper(WhereRel.Or, opers);
  }

  private static AndOrOper(whereRel: WhereRel, opers: Cond[]): Logic {
    var res = new Logic();
    res.Clause = new WhereClause(Utils.ConvertWhereRel(whereRel));
    for (var i = 0; i < opers.length; i++) {
      res.Clause.Terms.push(opers[i]["Term"]);
    }
    return res;
  }

  private static AndOrLogic(whereRel: WhereRel, logics: Logic[]): Logic {
    var res = new Logic();
    res.Clause = new WhereClause(Utils.ConvertWhereRel(whereRel));
    for (var i = 0; i < logics.length; i++) {
      res.Clause.SubClauses.push(logics[i].Clause);
    }
    return res;
  }

}

File: Qb.Net/Logic.cs
using Viten.QueryBuilder.SqlOm;

namespace Viten.QueryBuilder
{

  /// <summary>Класс определения логических операций</summary>
  public class Logic
  {
    internal WhereClause Clause;

    internal Logic(WhereRel whereRel)
    {
      Clause = new WhereClause(whereRel);
    }

    /// <summary>Определение операции AND</summary>
    public static Logic And(params Cond[] opers)
    {
      return AndOr(WhereRel.And, opers);
    }

    /// <summary>Определение операции OR</summary>
    public static Logic Or(params Cond[] opers)
    {
      return AndOr(WhereRel.Or, opers);
    }

    static Logic AndOr(WhereRel relationship, params Cond[] opers)
    {
      Logic retVal = new Logic(relationship);
      if (opers != null)
        for (int i = 0; i < opers.Length; i++)
          retVal.Clause.Terms.Add(opers[i].Term);
      return retVal;
    }

    /// <summary>Определение операции AND</summary>
    public static Logic And(params Logic[] logics)
    {
      return AndOr(WhereRel.And, logics);
    }

    /// <summary>Определение операции OR</summary>
    public static Logic Or(params Logic[] logics)
    {
      return AndOr(WhereRel.Or, logics);
    }

    static Logic AndOr(WhereRel relationship, params Logic[] logics)
    {
      Logic retVal = new Logic(relationship);
      if (logics != null)
        for (int i = 0; i < logics.Length; i++)
          retVal.Clause.SubClauses.Add(logics[i].Clause);
      return retVal;
    }

  }
}
